providers/providers.go

```diff
@@ -126,52 +128,56 @@ func loadProvSet(dstore ds.Datastore, k cid.Cid) (*providerSet, error) {
 			continue
 		}
 
+		// check expiration time
+		t, err := readTimeValue(e.Value)
+		switch {
+		case err != nil:
+			// couldn't parse the time
+			log.Warning("parsing providers record from disk: ", err)
+			fallthrough
+		case now.Sub(t) > ProvideValidity:
+			// or just expired
+			err = dstore.Delete(ds.RawKey(e.Key))
+			if err != nil && err != ds.ErrNotFound {
+				log.Warning("failed to remove provider record from disk: ", err)
+			}
+			continue
+		}
+
 		lix := strings.LastIndex(e.Key, "/")
 
 		decstr, err := base32.RawStdEncoding.DecodeString(e.Key[lix+1:])
 		if err != nil {
 			log.Error("base32 decoding error: ", err)
+			err = dstore.Delete(ds.RawKey(e.Key))
+			if err != nil && err != ds.ErrNotFound {
+				log.Warning("failed to remove provider record from disk: ", err)
+			}
 			continue
 		}
 
 		pid := peer.ID(decstr)
 
-		t, err := readTimeValue(e.Value)
-		if err != nil {
-			log.Warning("parsing providers record from disk: ", err)
-			continue
-		}
-
 		out.setVal(pid, t)
 	}
 
 	return out, nil
 }
 
-func readTimeValue(i interface{}) (time.Time, error) {
-	data, ok := i.([]byte)
-	if !ok {
-		return time.Time{}, fmt.Errorf("data was not a []byte")
+func readTimeValue(data []byte) (time.Time, error) {
```

Comment:
avoiding the `interface{}` is a nice little gain! No more allocation for passing byte slices.

Reply:
(this was left over from the datastore refactor when we replaced all the interfaces with `[]byte`)